### PR TITLE
Use pow of two capacity for hash maps to allow for `& (n-1)` instead of `% n`

### PR DESCRIPTION
**Pull request by gingerBill**

### CODE REVIEW

File: core/runtime/dynamic_map_internal.odin [hunk 1]
```diff
@@ -300,17 +302,36 @@ __dynamic_map_set :: proc(h: Map_Header, hash: Map_Hash, value: rawptr, loc := #
...
-		// index = __dynamic_map_find(h, hash).entry_index
-		// assert(index >= 0)
 	}
 	
 	return __dynamic_map_get_entry(h, index)
 }
 
 
+@(private="file")
+next_pow2 :: proc "contextless" (n: int) -> int {
```
Comment (Kelimion) [inline core/runtime/dynamic_map_internal.odin:312]: This is a bit poorly named. If you pass it a power of two, it returns that same power of two, not the next one.

e.g.
```
next_pow2(001): 1
next_pow2(002): 2
next_pow2(003): 4
next_pow2(004): 4
next_pow2(005): 8
next_pow2(006): 8
next_pow2(007): 8
next_pow2(008): 8
next_pow2(009): 16
next_pow2(010): 16
next_pow2(011): 16
next_pow2(012): 16
next_pow2(013): 16
next_pow2(014): 16
next_pow2(015): 16
next_pow2(016): 16
next_pow2(017): 32
```

File: core/runtime/dynamic_map_internal.odin [hunk 2]
```diff
@@ -300,17 +302,36 @@ __dynamic_map_set :: proc(h: Map_Header, hash: Map_Hash, value: rawptr, loc := #
...
+	n := n
+	if n <= 0 {
+		return 0
+	} else if n <= 2 {
+		return n
+	}
+	n -= 1
```
Comment (Kelimion) [inline core/runtime/dynamic_map_internal.odin:319]: If replaced thusly, it does produce 16 for 8, 32 for 16, and so on:
```cpp
if intrinsics.count_ones(n) != 1 {
    n -= 1
}
```

**Review by Kelimion — COMMENTED** [review]
Looks good to me in principle, but does it doesn't seem to actually grow if the capacity is already a power of two, because `next_pow2` returns the same power of two you input. I suggested a fix.

File: core/runtime/dynamic_map_internal.odin (hunk 2 above)
Comment (Kelimion) [reply core/runtime/dynamic_map_internal.odin:319]: Oh, I see. The `next_pow2` is only for the initial reservation, and it just doubles after that. I withdraw my objections.

File: core/runtime/dynamic_map_internal.odin (hunk 1 above)
Comment (gingerBill) [reply core/runtime/dynamic_map_internal.odin:312]: Good test!

**Comment by gingerBill** [comment]
I forgot to say how much I love that this PR number was 2048, a power of two 😆

**Pull request merged**